apps/emqx_durable_storage/src/emqx_ds.erl

```diff
@@ -163,7 +163,14 @@
         %% shard (if applicable to the backend), as the batch will be split accordingly
         %% even if this flag is `true'.
         %% Default: `false'.
-        atomic => boolean()
+        atomic => boolean(),
+        %% Defines whether to automatically assign strictly monotonically increasing
+        %% internal timestamps when storing messages (which precludes upserting even if
+        %% crafting timestamps in the message), or allow repeated timestamps in the DB
+        %% (which allows upserting keys, depending on the properties of the layout
+        %% keymapper).
+        %% Default: `true'.
+        auto_assign_timestamps => boolean()
```

Comment:
Hm, I have to say I 100% thought that we agreed on the approach where this is part of batch entry rather than whole batch. Having it per-batch seemingly brings too much of unnecessary complexity.

I.e. something close to:
```
-spec store_batch(db(), [emqx_types:message() | {_TsID :: emqx_ds:time(), emqx_types:message()}], message_store_opts()) ->
    store_batch_result().
```

Reply:
> Hm, I have to say I 100% thought that we agreed on the approach where this is part of batch entry rather than whole batch.

Yes, and I thought that it implies having this option per-batch, so that we know whether to auto assign or not.

With the proposed spec change, wouldn't that move the auto-assignment to the caller instead of replication layer?   Or are you suggesting that, if `TsID` is defined by the caller, it is not subject to the auto-assign?

I didn't understand that suggestion when we talked before.  I'll adapt to use this then.

Reply:
Sounds like a cleaner approach indeed, thanks for the suggestion.  :smile_cat:

Reply:
:heavy_check_mark: